File: Ruby/Euler023.rb
#!/usr/bin/env ruby

def main()
  target = 28123
  abundants = abundant_numbers()
      .take_while { |x| x <= target }
      .to_a
  abundant_sums = abundants
      .combination(2).lazy
      .map { |pair| pair.sum }
      .filter { |x| x <= target }
      .uniq
      .to_a

  abundant_sums += abundants.map { |x| x * 2 }.take_while { |x| x <= target}

  sum_of_abundants = abundant_sums.lazy.uniq.sum
  sum_of_all = (target + 1) / 2 * target
  sum_of_non_abundants = sum_of_all - sum_of_abundants

  puts sum_of_non_abundants
end

def abundant_numbers()
  Enumerator.new do |enum|
    n = 1
    while true
      if is_abundant(n)
        enum.yield(n)
      end
      n += 1
    end
  end
end

def is_abundant(n)
  return get_factors(n).sum > n
end

def get_factors(of_int)
  max = Math.sqrt(of_int).ceil

  factors = (1..max).select { |factor| of_int % factor == 0 }
      .map { |factor| [factor, of_int / factor] }
      .flatten
      .filter { |factor| factor != of_int}
      .uniq

  return factors
end

if __FILE__ == $0
  main
end

File: Kotlin/Euler023.kt
package Euler023

import Combinatorics.combinationsWithReplacement

fun main() {
  val target = 28123
  val abundants = abundantNumbers().takeWhile { target >= it }.toList()

  val abundantSums = combinationsWithReplacement(abundants, 2)
    .map { it.sum() }
    .filter { target >= it }
    .distinct()

  // .sum() collapses the generator, so it eats a lot of memory because Java
  val sumOfAbundantSums = abundantSums.sum()
  val sumOfAll = (target + 1) / 2 * target
  val sumOfNonAbundants = sumOfAll - sumOfAbundantSums
  println(sumOfNonAbundants)
}

fun abundantNumbers(): Sequence<Int> {
  return generateSequence(1){it + 1}.asSequence().filter { isAbundant(it) }.asSequence()
}

fun isAbundant(n: Int): Boolean {
  return getFactors(n).sum() > n
}

fun getFactors(n: Int): Sequence<Int> {
  val seq = sequence {
    yield(1)
    val potentialFactorSequence = generateSequence(2) { it + 1 }.takeWhile { n >= it * it }
    for (i in potentialFactorSequence) {
      if (n % i == 0) {
        yield(i)
        if (i * i != n) {
          yield(n / i)
        }
      }
    }
  }
  return seq
}
